Catching up with TestKit

robsdedude opened the pull request:
1. Add support for routing table specific TestKit messaged (retrieving routing tables and forcing updates)
2. Improve logging in the TestKit backend (all logs-driver and bolt-got both to stdout and TestKit) for easier debugging
3. Overhaul how/when the driver drops servers from the cached routing table:
   * Move logic of deactivating servers on failures into the pool. This functionality should not only be present when using transaction functions.
   * Drop writers on certain error codes: `Neo.ClientError.Cluster.NotALeader` and `Neo.ClientError.General.ForbiddenOnReadOnlyDatabase`
4. To simplify the code and avoid deadlocks or inconsistent driver states, both pool and router use blocking locks now. However, they will never perform IO while holding the lock removing the need for lock acquisition timeouts.

Depends on:
 * https://github.com/neo4j-drivers/testkit/pull/587

robsdedude committed:
Reformat feature list to reflect TestKit's state

robsdedude committed:
Clean up test skips

robsdedude committed:
Go driver does not support liveness check

robsdedude committed:
TestKit backend: improve logging

Log everything to stdout + TestKit socket.

robsdedude committed:
Drop servers from routing table on IO failure

robsdedude committed:
Remove writer from routing table on certain errors

robsdedude committed:
Refactor responsibilities and locking

The retry state is no longer responsible for invalidating servers on broken
connections. The pool and/or routing logic should take care of that. This is
necessary and logical since invalidation should happen regardless of which API
(session.Run, transaction.Run, ...) is used.

Pool and router no longer do any IO while holding locks. Hence, the locks can be
turned into blocking locks without risking blocking for too long.

robsdedude committed:
Add support for RT related TestKit messages

robsdedude committed:
Merge branch '5.0' into testkit-update-backend

fbiville committed:
Remove unused context parameters

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

robsdedude committed:
Remove more unused ctx parameters

StephenCathcart commented on neo4j/driver_with_context.go:
```diff
@@ -394,9 +397,7 @@ func (d *driverWithContext) Close(ctx context.Context) error {
 	// Safeguard against closing more than once
 	if d.pool != nil {
-		if err := d.pool.Close(ctx); err != nil {
-			return err
```
Is there a reason we want to remove the error check of pool.Close()?

StephenCathcart replied:
Ignore my comment - func has removed error return 👍

robsdedude committed:
Remove redundant Mutex.Unlock call

StephenCathcart reviewed (approved):
LGTM! :shipit:

robsdedude committed:
Merge branch '5.0' into testkit-update-backend

fbiville commented on neo4j/internal/bolt/bolt5.go:
```diff
@@ -921,19 +922,23 @@ func (b *bolt5) reAuth(ctx context.Context, auth *idb.ReAuthToken) error {
 	b.databaseName = database
 }
 
+func (b *bolt5) SelectedDatabase() string {
```
nit: I'd personally go for just `Database()`

fbiville commented on neo4j/internal/bolt/connections.go:
```diff
@@ -27,7 +27,11 @@ import (
 
-type Neo4jErrorCallback func(context.Context, idb.Connection, *db.Neo4jError) error
+type ConnectionErrorListener interface {
+	OnNeo4jError(context.Context, idb.Connection, *db.Neo4jError) error
```
why do they return errors?

robsdedude replied:
`pool.OnNeo4jError` might call the auth manager which is fallible.

The other methods used to be fallible because of the racing locks they had. They don't any longer so this could be changed. I'm just wondering how valid it is to assume that the pool might want to start failing in those functions in the future. Not sure. What do you think?

fbiville replied:
Those are internal, so no need to be future-proof, I'd say. If they don't return errors, let's just remove these from the signatures?

robsdedude replied:
:zap:

fbiville commented on testkit-backend/backend.go:
```diff
@@ -150,6 +150,7 @@ func (b *backend) writeLine(s string) error {
 func (b *backend) writeLineLocked(s string) error {
 	b.wrLock.Lock()
 	defer b.wrLock.Unlock()
+	fmt.Println(s)
```
I think this was added on purpose, but I forgot why. Can you tell (again)?

robsdedude replied:
So that log output goes to both sides: driver (this `Println`) and TestKit (`return b.writeLine(s)` in the next line).

robsdedude committed:
Fix not removing server from all routing tables

Also removing error return value from all router.Invalidate* calls

robsdedude committed:
TestKit backend: routing table response

TestKit expects the default database to be `null` not an empty string.
So we need to translate.

robsdedude committed:
Add (back) test skips

fbiville committed:
Rename `SelectedDatabase()` -> `Database()`

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

fbiville committed:
Remove error return value from internal infallible interface functions

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

robsdedude committed:
Fix silly copy paste error
